### PR TITLE
Check ansible playbook after job schedule...

### PR DESCRIPTION
**Pull request by amolpati30**
- After adding a new user and running the job, the system displayed the root user instead of the new user. 

- The automation test below is written to verify that the new user is correctly set after the job runs.

### CODE REVIEW

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_privilege_escalation_ansible_playbook

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_privilege_escalation_ansible_playbook

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 7902
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_privilege_escalation_ansible_playbook --external-logging
Test Result : ========== 1 failed, 11 deselected, 24 warnings in 1181.28s (0:19:41) ==========
```

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_privilege_escalation_ansible_playbook

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 7910
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_privilege_escalation_ansible_playbook --external-logging
Test Result : ========== 1 passed, 11 deselected, 24 warnings in 969.20s (0:16:09) ===========
```

File: tests/foreman/api/test_ansible.py [hunk 1]
```diff
@@ -602,3 +602,102 @@ def test_negative_ansible_job_timeout_to_kill(
...
+        assert user1.status == 0
+        echo1 = rhel_contenthost.execute('echo dog8code | passwd rexuser --stdin')
+        assert 'passwd: all authentication tokens updated successfully.' in echo1.stdout
+        pswd = rhel_contenthost.execute(
+            'echo "rexuser   ALL=NOPASSWD:   ALL" | tee -a /etc/sudoers.d/rexuser'
+        )
+        assert pswd.status == 0
+        user2 = rhel_contenthost.execute('useradd testing')
+        assert user2.status == 0
+        echo2 = rhel_contenthost.execute('echo dog8code | passwd testing --stdin')
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_ansible.py:667]:
```suggestion
        rex_passwd = rhel_contenthost.execute(f'echo {settings.server.ssh_password} | passwd testing --stdin')
```

File: tests/foreman/api/test_ansible.py [hunk 2]
```diff
@@ -602,3 +602,102 @@ def test_negative_ansible_job_timeout_to_kill(
...
+                        - debug:
+                            var: bec_user.stdout
+                '''
+        result = rhel_contenthost.register(
+            module_org, module_location, module_activation_key.name, target_sat
+        )
+        assert result.status == 0, f'Failed to register host: {result.stderr}'
+        assert rhel_contenthost.username == 'root'
+        user1 = rhel_contenthost.execute('useradd rexuser')
+        assert user1.status == 0
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_ansible.py:658]:
```suggestion
        assert rhel_contenthost.execute('useradd rexuser').status == 0
```

File: tests/foreman/api/test_ansible.py [hunk 3]
```diff
@@ -602,3 +602,102 @@ def test_negative_ansible_job_timeout_to_kill(
...
+                          become_user: testing
+                          register: bec_user
+                        - debug:
+                            var: bec_user.stdout
+                '''
+        result = rhel_contenthost.register(
+            module_org, module_location, module_activation_key.name, target_sat
+        )
+        assert result.status == 0, f'Failed to register host: {result.stderr}'
+        assert rhel_contenthost.username == 'root'
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_ansible.py:656]: what this line is for?

File: tests/foreman/api/test_ansible.py [hunk 4]
```diff
@@ -602,3 +602,102 @@ def test_negative_ansible_job_timeout_to_kill(
...
+        assert pswd.status == 0
+        user2 = rhel_contenthost.execute('useradd testing')
+        assert user2.status == 0
+        echo2 = rhel_contenthost.execute('echo dog8code | passwd testing --stdin')
+        assert 'passwd: all authentication tokens updated successfully.' in echo2.stdout
+        whoami = rhel_contenthost.execute('su - testing -c "whoami"')
+        assert 'testing' in whoami.stdout or whoami.status == 0
+        setting_update.value = ''
+        setting_update = setting_update.update({'value'})
+        assert setting_update.value == ''
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_ansible.py:673]: Instead of updating remote_execution_effective_user setting at satellite, why not using effective_user from the REX job create page?

File: tests/foreman/api/test_ansible.py [hunk 5]
```diff
@@ -602,3 +602,102 @@ def test_negative_ansible_job_timeout_to_kill(
...
+        pswd = rhel_contenthost.execute(
+            'echo "rexuser   ALL=NOPASSWD:   ALL" | tee -a /etc/sudoers.d/rexuser'
+        )
+        assert pswd.status == 0
+        user2 = rhel_contenthost.execute('useradd testing')
+        assert user2.status == 0
+        echo2 = rhel_contenthost.execute('echo dog8code | passwd testing --stdin')
+        assert 'passwd: all authentication tokens updated successfully.' in echo2.stdout
+        whoami = rhel_contenthost.execute('su - testing -c "whoami"')
+        assert 'testing' in whoami.stdout or whoami.status == 0
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_ansible.py:670]: just wondering, why verifying whoami command here?  looks unrelated

File: tests/foreman/api/test_ansible.py [hunk 6]
```diff
@@ -602,3 +602,102 @@ def test_negative_ansible_job_timeout_to_kill(
...
+                        - debug:
+                            var: def_user.stdout
+                        - name: Check become user
+                          command: bash -c "whoami"
+                          become: true
+                          become_user: testing
+                          register: bec_user
+                        - debug:
+                            var: bec_user.stdout
+                '''
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_ansible.py:651]:
```suggestion
        playbook = '''
            ---
            - name: Test Play
              hosts: all
              gather_facts: false
              tasks:
                - name: Check current user
                    command: bash -c "whoami"
                    register: def_user
                - debug:
                    var: def_user.stdout
                - name: Check become user
                    command: bash -c "whoami"
                    become: true
                    become_user: testing
                    register: bec_user
                - debug:
                    var: bec_user.stdout
        '''
```

File: tests/foreman/api/test_ansible.py [hunk 7]
```diff
@@ -602,3 +602,102 @@ def test_negative_ansible_job_timeout_to_kill(
         assert [i['output'] for i in result if i['output'] == termination_msg]
         assert [i['output'] for i in result if i['output'] == 'StandardError: Job execution failed']
         assert [i['output'] for i in result if i['output'] == 'Exit status: 120']
+
+    @pytest.mark.tier2
+    @pytest.mark.no_containers
+    @pytest.mark.rhel_ver_match('8')
+    @pytest.mark.parametrize('setting_update', ['remote_execution_effective_user'], indirect=True)
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_ansible.py:609]: It won't be required if we handle effective user at job level, right?

File: tests/foreman/api/test_ansible.py [hunk 8]
```diff
@@ -602,3 +602,102 @@ def test_negative_ansible_job_timeout_to_kill(
...
+        :id: 8c63fd1a-2121-4cce-9ec1-ae12817c9cc4
+
+        :steps:
+            1. Register a RHEL host to Satellite.
+            2. Setup a user on that host.
+            3. Create a playbook.
+            4. Set the SSH user to the created user, and unset the Effective user.
+            5. Run the playbook.
+
+        :expectedresults: In the playbook, created user is expected instead root user.
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_ansible.py:631]: Could you add related jira/bz and customerscenario: true for this bug?

File: tests/foreman/api/test_ansible.py (hunk 1 above)
Comment (amolpati30) [reply tests/foreman/api/test_ansible.py:667]: updated changes for password.

File: tests/foreman/api/test_ansible.py (hunk 4 above)
Comment (amolpati30) [reply tests/foreman/api/test_ansible.py:673]: There is no option available in Nailgun to update the SSH user remotely. So using `setting_update` updating the value.

File: tests/foreman/api/test_ansible.py (hunk 6 above)
Comment (amolpati30) [reply tests/foreman/api/test_ansible.py:651]: The required changes have been made.

File: tests/foreman/api/test_ansible.py (hunk 7 above)
Comment (amolpati30) [reply tests/foreman/api/test_ansible.py:609]: I am handling it by updating the settings page. The JobInvocation() method does not have an option to update the effective user.

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_privilege_escalation_ansible_playbook

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 7922
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_privilege_escalation_ansible_playbook --external-logging
Test Result : ========== 1 passed, 11 deselected, 24 warnings in 980.51s (0:16:20) ===========
```

File: tests/foreman/api/test_ansible.py (hunk 4 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_ansible.py:677]: yes, effective_user is not available but you can add it anyway, its already part of apidoc

File: tests/foreman/api/test_ansible.py (hunk 7 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_ansible.py:609]: you can add it in the job invocation method then

File: tests/foreman/api/test_ansible.py (hunk 4 above)
Comment (amolpati30) [reply tests/foreman/api/test_ansible.py:677]: I tried it before, but it didn't give the proper output for two users.

File: tests/foreman/api/test_ansible.py (hunk 7 above)
Comment (amolpati30) [reply tests/foreman/api/test_ansible.py:609]: "I can add that, but as mentioned above, it doesn't produce valid output.

File: tests/foreman/api/test_ansible.py [hunk 9]
```diff
@@ -602,3 +602,106 @@ def test_negative_ansible_job_timeout_to_kill(
...
+        self,
+        target_sat,
+        rhel_contenthost,
+        module_org,
+        module_location,
+        module_ak_with_synced_repo,
+        setting_update,
+    ):
+        """Privilege escalation defined inside ansible playbook tasks is working
+        when executing the playbook via Remote Execution
```
[the remark this answers is not in the document]
Comment (amolpati30) [reply tests/foreman/api/test_ansible.py]: Changes has been done.

File: tests/foreman/api/test_ansible.py [hunk 10]
```diff
@@ -602,3 +602,106 @@ def test_negative_ansible_job_timeout_to_kill(
         assert [i['output'] for i in result if i['output'] == termination_msg]
         assert [i['output'] for i in result if i['output'] == 'StandardError: Job execution failed']
         assert [i['output'] for i in result if i['output'] == 'Exit status: 120']
+
+    @pytest.mark.tier2
+    @pytest.mark.no_containers
+    @pytest.mark.rhel_ver_match('8')
+    @pytest.mark.parametrize('setting_update', ['remote_execution_effective_user'], indirect=True)
+    def test_positive_privilege_escalation_ansible_playbook(
```
[the remark this answers is not in the document]
Comment (amolpati30) [reply tests/foreman/api/test_ansible.py]: Changed the name of the test_case as per suggestion

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_ansible_job_privilege_escalation

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8064
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_ansible_job_privilege_escalation --external-logging
Test Result : ========== 1 passed, 11 deselected, 24 warnings in 1053.58s (0:17:33) ==========
```

File: tests/foreman/api/test_ansible.py [hunk 11]
```diff
@@ -602,3 +602,106 @@ def test_negative_ansible_job_timeout_to_kill(
...
+        assert result.status == 0, f'Failed to register host: {result.stderr}'
+        assert rhel_contenthost.execute('useradd rexuser').status == 0
+        rex_passwd1 = rhel_contenthost.execute(
+            f'echo {settings.server.ssh_password} | passwd rexuser --stdin'
+        )
+        assert 'passwd: all authentication tokens updated successfully.' in rex_passwd1.stdout
+        pswd = rhel_contenthost.execute(
+            'echo "rexuser   ALL=NOPASSWD:   ALL" | tee -a /etc/sudoers.d/rexuser'
+        )
+        assert pswd.status == 0
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_ansible.py:668]: It seems we're not using rexuser at all, can we remove it?

File: tests/foreman/api/test_ansible.py [hunk 12]
```diff
@@ -602,3 +602,106 @@ def test_negative_ansible_job_timeout_to_kill(
...
+        assert rhel_contenthost.execute('useradd testing').status == 0
+        rex_passwd2 = rhel_contenthost.execute(
+            f'echo {settings.server.ssh_password} | passwd testing --stdin'
+        )
+        assert 'passwd: all authentication tokens updated successfully.' in rex_passwd2.stdout
+        rhel_contenthost.execute('su - testing -c "whoami"')
+        setting_update.value = ''
+        setting_update = setting_update.update({'value'})
+        assert setting_update.value == ''
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_ansible.py:677]: this looks redundant, can we remove this lines?

File: tests/foreman/api/test_ansible.py (hunk 7 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_ansible.py:609]: As discussed and as per BZ, we can skip setting effective_user to run this playbook

File: tests/foreman/api/test_ansible.py (hunk 11 above)
Comment (amolpati30) [reply tests/foreman/api/test_ansible.py]: Yes, it is not required, so I am removing it.

File: tests/foreman/api/test_ansible.py (hunk 12 above)
Comment (amolpati30) [reply tests/foreman/api/test_ansible.py]: I've removed redundant lines.

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_ansible_job_privilege_escalation

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8078
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_ansible_job_privilege_escalation --external-logging
Test Result : ========== 1 passed, 11 deselected, 24 warnings in 2020.15s (0:33:40) ==========
```

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_ansible_job_privilege_escalation

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8081
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/api/test_ansible.py::TestAnsibleREX -k test_positive_ansible_job_privilege_escalation --external-logging
Test Result : ========== 1 passed, 11 deselected, 24 warnings in 956.92s (0:15:56) ===========
```

**Review by Gauravtalreja1 — APPROVED** [review]
